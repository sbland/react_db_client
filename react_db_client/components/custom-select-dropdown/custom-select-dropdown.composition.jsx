import React, { useState } from 'react';
import { CustomSelectDropdown } from './custom-select-dropdown';


const DEMO_OPTIONS = [
  { uid: 1, label: '01' },
  { uid: 2, label: '02' },
  { uid: 3, label: '0ssdfsadfasdfd ldj ldjlkd jldjk 3' },
  { uid: 4, label: '04' },
  { uid: 5, label: '0asdfsf5' },
  { uid: 6, label: '06' },
  { uid: 7, label: '07' },
];
const DEMO_OPTIONS_LONG = [...Array(30)].map((v, i) => ({ uid: i, label: i }));

export const Basic = () => {
  return (
    <CustomSelectDropdown
      options={DEMO_OPTIONS}
      handleSelect={(uid) => {}}
      isOpen
      handleClose={() => {}}
      firstItemRef={{ current: {} }}
      goBackToSearchField={() => {}}
    />
  )
}

export const Interactive = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState(null);
  return (
    <div style={{ width: '10rem' }}>
      {selection || 'Nothing selected'}
      <input
        type="text"
        onFocus={() => setOpen(true)}
        style={{ width: '100%' }}
      />
      <CustomSelectDropdown
        options={DEMO_OPTIONS}
        handleSelect={(uid) => setSelection(uid)}
        isOpen={open}
        handleClose={() => setOpen(false)}
        firstItemRef={{ current: {} }}
      />
      <div style={{ width: '300px', height: '300px', background: 'red' }} />
    </div>
  );
};

export const LongList = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState(null);
  return (
    <div style={{ width: '10rem' }}>
      {selection || 'Nothing Selected'}
      <input
        type="text"
        onFocus={() => setOpen(true)}
        style={{ width: '100%' }}
      />
      <CustomSelectDropdown
        options={DEMO_OPTIONS_LONG}
        handleSelect={(uid) => setSelection(uid)}
        isOpen={open}
        handleClose={() => setOpen(false)}
        firstItemRef={{ current: {} }}
      />
      <div style={{ width: '300px', height: '300px', background: 'red' }} />
    </div>
  );
};

export const OverflowHidden = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState(null);
  // TODO: This is currently not working
  return (
    <div
      style={{
        width: '10rem',
        height: '5rem',
        overflow: 'hidden',
        position: 'relative',
      }}
    >
      {selection || 'Nothing Selected'}
      <input
        type="text"
        onFocus={() => setOpen(true)}
        style={{ width: '100%' }}
      />
      <div style={{ position: 'absolute', zIndex: 99 }}>
        <CustomSelectDropdown
          options={DEMO_OPTIONS_LONG}
          handleSelect={(uid) => setSelection(uid)}
          isOpen={open}
          handleClose={() => setOpen(false)}
          firstItemRef={{ current: {} }}
        />
      </div>
      <div style={{ width: '300px', height: '300px', background: 'red' }} />
    </div>
  );
};
