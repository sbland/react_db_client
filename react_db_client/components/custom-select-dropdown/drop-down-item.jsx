import React from 'react';
import PropTypes from 'prop-types';

export const DropDownItem = ({ itemRef, uid, label, handleSelect, isSelected, handleKeyDown }) => (
  <li className="selectDropdown_item">
    <button
      type="button"
      className={`itemBtn ${isSelected ? 'button-two selected' : 'button-one notSelected'}`}
      onClick={() => handleSelect(uid)}
      ref={itemRef}
      onKeyDown={handleKeyDown}
    >
      {label || 'Missing Label'}
    </button>
  </li>
);

DropDownItem.propTypes = {
  itemRef: PropTypes.func,
  uid: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
  handleKeyDown: PropTypes.func.isRequired,
};

DropDownItem.defaultProps = {
  itemRef: null,
  isSelected: false,
};
