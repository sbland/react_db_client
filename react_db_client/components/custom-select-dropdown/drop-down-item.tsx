import React from 'react';
import PropTypes from 'prop-types';
import { SelectDropwdownItem } from './styles';

export const DropDownItem = ({ itemRef, uid, label, handleSelect, isSelected, handleKeyDown }) => (
  <SelectDropwdownItem>
    <button
      type="button"
      className={`itemBtn ${isSelected ? 'selected' : 'notSelected'}`}
      onClick={() => handleSelect(uid)}
      ref={itemRef}
      onKeyDown={handleKeyDown}
    >
      {label || 'Missing Label'}
    </button>
  </SelectDropwdownItem>
);

DropDownItem.propTypes = {
  itemRef: PropTypes.func,
  uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
  handleKeyDown: PropTypes.func.isRequired,
};

DropDownItem.defaultProps = {
  itemRef: null,
  isSelected: false,
};
