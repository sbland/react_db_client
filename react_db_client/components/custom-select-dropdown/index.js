export { CustomSelectDropdown } from './custom-select-dropdown';
export { DropDownItem } from './drop-down-item';
