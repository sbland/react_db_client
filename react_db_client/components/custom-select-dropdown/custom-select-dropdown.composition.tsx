import React, { useState } from 'react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { CustomSelectDropdown } from './custom-select-dropdown';

const DEMO_OPTIONS = [
  { uid: 1, label: '01' },
  { uid: 2, label: '02' },
  { uid: 3, label: '0ssdfsadfasdfd ldj ldjlkd jldjk 3' },
  { uid: 4, label: '04' },
  { uid: 5, label: '0asdfsf5' },
  { uid: 6, label: '06' },
  { uid: 7, label: '07' },
];
const DEMO_OPTIONS_LONG = [...Array(30)].map((v, i) => ({ uid: i, label: String(i) }));

export const Basic = () => {
  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      <CustomSelectDropdown
        options={DEMO_OPTIONS}
        handleSelect={(uid) => {}}
        isOpen
        handleClose={() => {}}
        firstItemRef={{ current: {} as HTMLElement }}
        goBackToSearchField={() => {}}
      />
    </CompositionWrapDefault>
  );
};

export const Interactive = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState<string | null>(null);
  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      <div style={{ width: '10rem' }}>
        <p data-testid="curSel">{selection || 'Nothing Selected'}</p>
        <input
          data-testid="testInput"
          type="text"
          onFocus={() => setOpen(true)}
          style={{ width: '100%' }}
        />
        <CustomSelectDropdown
          options={DEMO_OPTIONS}
          handleSelect={(uid) => setSelection(uid)}
          isOpen={open}
          handleClose={() => setOpen(false)}
          firstItemRef={{ current: {} as HTMLElement }}
          goBackToSearchField={() => {}}
        />
        <div style={{ width: '300px', height: '300px', background: 'red' }} />
      </div>
    </CompositionWrapDefault>
  );
};

export const LongList = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState<string | null>(null);
  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      <div style={{ width: '10rem' }}>
        {selection || 'Nothing Selected'}
        <input type="text" onFocus={() => setOpen(true)} style={{ width: '100%' }} />
        <CustomSelectDropdown
          options={DEMO_OPTIONS_LONG}
          handleSelect={(uid) => setSelection(uid)}
          isOpen={open}
          handleClose={() => setOpen(false)}
          firstItemRef={{ current: {} as HTMLElement }}
          goBackToSearchField={() => {}}
        />
        <div style={{ width: '300px', height: '300px', background: 'red' }} />
      </div>
    </CompositionWrapDefault>
  );
};

export const OverflowHidden = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState<string | null>(null);
  // TODO: This is currently not working
  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      <div
        style={{
          width: '10rem',
          height: '5rem',
          overflow: 'hidden',
          position: 'relative',
        }}
      >
        <p data-testid="curSel">{selection || 'Nothing Selected'}</p>
        <input type="text" onFocus={() => setOpen(true)} style={{ width: '100%' }} />
        <div style={{ position: 'absolute', zIndex: 99 }}>
          <CustomSelectDropdown
            options={DEMO_OPTIONS_LONG}
            handleSelect={(uid) => setSelection(uid)}
            isOpen={open}
            handleClose={() => setOpen(false)}
            firstItemRef={{ current: {} as HTMLElement }}
            goBackToSearchField={() => {}}
          />
        </div>
        <div style={{ width: '300px', height: '300px', background: 'red' }} />
      </div>
    </CompositionWrapDefault>
  );
};

export const OverflowHiddenFixed = () => {
  const [open, setOpen] = useState(false);
  const [selection, setSelection] = useState<string | null>(null);
  const containerRef = React.useRef(null);
  const targetRef = React.useRef<HTMLDivElement>(null);
  const [rerender, setRerender] = React.useState(0);
  const [position, setPosition] = React.useState({ top: 0, left: 0 });

  React.useEffect(() => {
    setRerender(1);
    if (targetRef.current) {
      setPosition({
        top: targetRef.current?.offsetTop,
        left: targetRef.current?.offsetLeft,
      });
    }
  }, [targetRef.current]);

  // TODO: This is currently not working
  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      <div
        style={{
          width: '10rem',
          height: '5rem',
          overflow: 'hidden',
          position: 'relative',
        }}
      >
        {selection || 'Nothing Selected'}
        <input type="text" onFocus={() => setOpen(true)} style={{ width: '100%' }} />
        <div style={{ position: 'absolute', zIndex: 99 }} ref={targetRef}>
          <CustomSelectDropdown
            options={DEMO_OPTIONS_LONG}
            handleSelect={(uid) => setSelection(uid)}
            isOpen={open}
            handleClose={() => setOpen(false)}
            firstItemRef={{ current: {} as HTMLElement }}
            position="absolute"
            absolutePosition={position}
            containerRef={containerRef.current}
            goBackToSearchField={() => {}}
          />
        </div>
        <div style={{ width: '300px', height: '300px', background: 'red' }} />
      </div>
      <div
        style={{ position: 'absolute', left: 0, top: 0, background: undefined }}
        ref={containerRef}
      >
        {' '}
      </div>
    </CompositionWrapDefault>
  );
};

export const BasicMenuAbsolutePosition = () => {
  const containerRef = React.useRef<HTMLDivElement>(null);
  const targetRef = React.useRef(null);
  const [rerender, setRerender] = React.useState(0);

  React.useEffect(() => {
    setRerender(1);
  }, []);

  return (
    <CompositionWrapDefault height="8rem" width="12rem">
      {(rerender && containerRef.current && (
        <CustomSelectDropdown
          options={DEMO_OPTIONS}
          handleSelect={(uid) => {}}
          isOpen
          handleClose={() => {}}
          firstItemRef={{ current: {} as HTMLElement }}
          goBackToSearchField={() => {}}
          containerRef={containerRef.current}
          position="absolute"
          absolutePosition={{ left: 50, top: 50 }}
        />
      )) || <>""</>}
      <div
        style={{ position: 'absolute', left: 10, top: 400, background: 'red' }}
        ref={containerRef}
      >
        Here!
      </div>

      <div
        style={{ position: 'absolute', left: 100, top: 400, background: 'green' }}
        ref={targetRef}
      >
        Here!
      </div>
    </CompositionWrapDefault>
  );
};
