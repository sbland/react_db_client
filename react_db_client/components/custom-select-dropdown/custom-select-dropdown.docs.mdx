---
description: 'A CustomSelectDropdown component.'
labels: ['ui', 'form', 'input']
---

import { CustomSelectDropdown } from './custom-select-dropdown';

## React Selection component

Custom selection component
