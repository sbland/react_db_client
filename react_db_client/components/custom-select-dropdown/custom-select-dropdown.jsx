import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import './customSelectDropdown.scss';
import { DropDownItem } from './drop-down-item';

export const CustomSelectDropdown = ({
  options,
  handleSelect,
  isOpen,
  handleClose,
  firstItemRef,
  goBackToSearchField = () => {},
}) => {
  return <div>Hello</div>
  const menuRef = useRef(null);
  const itemRefs = useRef([]);
  const [currentItemFocus, setCurrentItemFocus] = useState(0);

  useEffect(() => {
    itemRefs.current = itemRefs.current.slice(0, options.length);
  }, [options]);

  function handleClickOutside(event) {
    if (isOpen && menuRef.current && !menuRef.current.contains(event.target)) {
      handleClose();
    }
  }

  // TODO: Disabled as catches user input
  // useEffect(() => {
  //   if (isOpen) {
  //     // focus on first item when dropdown opened
  //     if (itemRefs.current[0]) itemRefs.current[0].focus();
  //   }
  // }, [isOpen]);

  useEffect(() => {
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  });

  const handleItemKeyDown = (e) => {
    if (e.key === 'ArrowDown' && currentItemFocus < options.length - 1) {
      e.preventDefault();
      itemRefs.current[currentItemFocus + 1].focus();
      setCurrentItemFocus((prev) => prev + 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus > 0) {
      e.preventDefault();
      itemRefs.current[currentItemFocus - 1].focus();
      setCurrentItemFocus((prev) => prev - 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus === 0) {
      e.preventDefault();
      goBackToSearchField();
      setCurrentItemFocus(0);
    }
    if (e.key === 'Escape') goBackToSearchField();
  };

  const mapOptions = options.map((opt, i) => (
    <DropDownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
      itemRef={(el) => {
        itemRefs.current[i] = el;
        // eslint-disable-next-line no-param-reassign
        if (i === 0) firstItemRef.current = el;
      }}
      handleKeyDown={(e) => handleItemKeyDown(e)}
      onFocus={() => setCurrentItemFocus(i)}
      isFocused={i === currentItemFocus}
    />
  ));
  // const mapOptions = <div>OPTIONS</div>

  const listDropdownClass = isOpen ? 'selectDropdown_list open' : 'selectDropdown_list';
  const menuDropdownClass = isOpen ? 'selectDropdown_menu open' : 'selectDropdown_menu';

  return (
    <div className="selectDropdown">
      <div className={menuDropdownClass} ref={menuRef}>
        {isOpen && <ul className={listDropdownClass}>{mapOptions}</ul>}
      </div>
    </div>
  );
};

CustomSelectDropdown.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    })
  ).isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  firstItemRef: PropTypes.object.isRequired,
  goBackToSearchField: PropTypes.func.isRequired,
};

