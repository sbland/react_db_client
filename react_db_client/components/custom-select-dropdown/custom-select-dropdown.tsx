import React, { useEffect, useRef, useState } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import { DropDownItem } from './drop-down-item';
import { SelectDropdown, SelectDropdownList, SelectDropdownMenu } from './styles';

export interface Option {
  uid: string | number;
  label: string;
}

export interface ICustomSelectDropdownProps {
  options: Option[];
  handleSelect: (uid: string) => void;
  isOpen: boolean;
  handleClose: () => void;
  firstItemRef: React.MutableRefObject<HTMLElement | null>;
  goBackToSearchField: () => void;
  containerRef?: HTMLElement | null;
  position?: 'absolute' | 'relative';
  absolutePosition?: { left: number; top: number } | null;
}

export const CustomSelectDropdown: React.FC<ICustomSelectDropdownProps> = ({
  options,
  handleSelect,
  isOpen,
  handleClose,
  firstItemRef,
  goBackToSearchField = () => null,
  containerRef = null,
  position = 'relative',
  absolutePosition = null,
}: ICustomSelectDropdownProps) => {
  const menuRef = useRef<HTMLDivElement>(null);
  const itemRefs = useRef<HTMLElement[]>([]);
  const [currentItemFocus, setCurrentItemFocus] = useState(0);

  useEffect(() => {
    itemRefs.current = itemRefs.current.slice(0, options.length);
  }, [options]);

  function handleClickOutside(event) {
    if (menuRef.current && !menuRef.current.contains(event.target)) {
      handleClose();
    }
  }

  useEffect(() => {
    // Bind the event listener
    if (isOpen) {
      document.addEventListener('mousedown', handleClickOutside);
    } else {
      document.removeEventListener('mousedown', handleClickOutside);
    }
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [isOpen]);

  const handleItemKeyDown = (e) => {
    if (e.key === 'ArrowDown' && currentItemFocus < options.length - 1) {
      e.preventDefault();
      itemRefs.current[currentItemFocus + 1].focus();
      setCurrentItemFocus((prev) => prev + 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus > 0) {
      e.preventDefault();
      itemRefs.current[currentItemFocus - 1].focus();
      setCurrentItemFocus((prev) => prev - 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus === 0) {
      e.preventDefault();
      goBackToSearchField();
      setCurrentItemFocus(0);
    }
    if (e.key === 'Escape') goBackToSearchField();
    if (e.key === 'Tab') {
      e.preventDefault();
      goBackToSearchField();
    }
  };

  const mapOptions = options.map((opt, i) => (
    <DropDownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
      itemRef={(el) => {
        itemRefs.current[i] = el;
        // eslint-disable-next-line no-param-reassign
        if (i === 0) firstItemRef.current = el;
      }}
      handleKeyDown={(e) => handleItemKeyDown(e)}
      // onFocus={() => setCurrentItemFocus(i)}
      // isFocused={i === currentItemFocus}
    />
  ));
  // const mapOptions = <div>OPTIONS</div>

  const listDropdownClass = isOpen ? 'selectDropdown_list open' : 'selectDropdown_list';
  const menuDropdownClass = isOpen ? 'selectDropdown_menu open' : 'selectDropdown_menu';

  const render = (
    <SelectDropdown
      className="selectDropdown"
      style={{ position, left: absolutePosition?.left || 0, top: absolutePosition?.top || 0 }}
    >
      <SelectDropdownMenu className={menuDropdownClass} ref={menuRef}>
        {isOpen && (
          <SelectDropdownList className={listDropdownClass}>{mapOptions}</SelectDropdownList>
        )}
      </SelectDropdownMenu>
    </SelectDropdown>
  );
  if (containerRef) return ReactDOM.createPortal(render, containerRef);
  return render;
};

CustomSelectDropdown.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    })
  ).isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  firstItemRef: PropTypes.object.isRequired,
  goBackToSearchField: PropTypes.func.isRequired,
  containerRef: PropTypes.any,
};

CustomSelectDropdown.defaultProps = {
  containerRef: null,
};
