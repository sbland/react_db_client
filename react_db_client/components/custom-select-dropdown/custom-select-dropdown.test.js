import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { MockReactC } from '@react_db_client/testing.utils';

import { CustomSelectDropdown } from './custom-select-dropdown';
import { DropDownItem } from './drop-down-item';

jest.mock('./drop-down-item', () => MockReactC('DropDownItem', ['DropDownItem']));

jest.mock('react', () => {
  const originReact = jest.requireActual('react');
  let i = 0;
  return {
    ...originReact,
    useRef: jest.fn().mockImplementation(() => {
      i += 1;
      if (i % 2 === 0)
        return { current: [{ focus: () => {} }, { focus: () => {} }, { focus: () => {} }] };
      return { current: null };
    }),
  };
});

const DEMO_OPTIONS = [
  { uid: 1, label: '01' },
  { uid: 2, label: '02' },
  { uid: 3, label: '03' },
  { uid: 4, label: '04' },
  { uid: 5, label: '05' },
  { uid: 6, label: '06' },
  { uid: 7, label: '07' },
];

const handleSelect = jest.fn();
const handleClose = jest.fn();
const goBackToSearchField = jest.fn();

const defaultProps = {
  options: DEMO_OPTIONS,
  handleSelect,
  isOpen: true,
  handleClose,
  firstItemRef: {}, // TODO: Set this correctly
  goBackToSearchField,
};

describe('CustomSelectDropdown', () => {
  beforeEach(() => {
    handleSelect.mockClear();
    handleClose.mockClear();
    goBackToSearchField.mockClear();
  });
  it('Renders', () => {
    shallow(<CustomSelectDropdown {...defaultProps} />);
  });
  it('Matches Snapshot', () => {
    const component = shallow(<CustomSelectDropdown {...defaultProps} />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('Unit Testing', () => {
    let component;
    beforeEach(() => {
      component = mount(<CustomSelectDropdown {...defaultProps} />);
    });

    describe('showing items', () => {
      test('should show a list of dropdown items', () => {
        expect(component.find(DropDownItem).length).toEqual(DEMO_OPTIONS.length);
      });
    });
    describe('selecting items', () => {
      test('should call handle select when one of the dropdown items is selected', () => {
        const item = component.find(DropDownItem).first();
        const itemUid = item.props().uid;
        item.props().handleSelect(itemUid);
        expect(handleSelect).toHaveBeenCalledWith(itemUid);
      });
    });
    describe('Navigating items', () => {
      test('should pass is focused prop to drop down item', () => {
        const item = component.find(DropDownItem).first();
        expect(item.props().isFocused).toEqual(true);
      });

      test('should allow scrolling through items with the up and down arrows', () => {
        let firstItem = component.find(DropDownItem).first();
        let secondItem = component.find(DropDownItem).at(1);
        expect(firstItem.props().isFocused).toEqual(true);
        expect(secondItem.props().isFocused).toEqual(false);
        act(() => {
          firstItem.props().handleKeyDown({ key: 'ArrowDown', preventDefault: () => {} });
        });
        component.update();
        firstItem = component.find(DropDownItem).first();
        secondItem = component.find(DropDownItem).at(1);
        expect(firstItem.props().isFocused).toEqual(false);
        expect(secondItem.props().isFocused).toEqual(true);
      });
    });
  });
});
