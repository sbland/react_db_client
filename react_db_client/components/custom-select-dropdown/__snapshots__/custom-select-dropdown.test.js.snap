// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSelectDropdown Matches Snapshot 1`] = `
"<div className=\\"selectDropdown\\">
  <div className=\\"selectDropdown_menu open\\">
    <ul className=\\"selectDropdown_list open\\">
      <DropDownItem uid={1} label=\\"01\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={true} />
      <DropDownItem uid={2} label=\\"02\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
      <DropDownItem uid={3} label=\\"03\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
      <DropDownItem uid={4} label=\\"04\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
      <DropDownItem uid={5} label=\\"05\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
      <DropDownItem uid={6} label=\\"06\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
      <DropDownItem uid={7} label=\\"07\\" handleSelect={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} itemRef={[Function: itemRef]} handleKeyDown={[Function: handleKeyDown]} onFocus={[Function: onFocus]} isFocused={false} />
    </ul>
  </div>
</div>"
`;
